src/doorman/Channel.cpp

```diff
@@ -0,0 +1,393 @@
+#include "doorman.hpp"
+
+#include <arpa/inet.h>
+#include <arpa/telnet.h>
+#include <csignal>
+#include <cstring>
+#include <netdb.h>
+
+using namespace std::literals;
+
+static constexpr auto MaxIncomingDataBufferSize = 2048u;
+
+static void SendCom(const Fd &fd, byte a, byte b) {
+    byte buf[] = {IAC, a, b};
+    fd.write(buf);
+}
+
+static void SendOpt(const Fd &fd, byte a) {
+    byte buf[] = {IAC, SB, a, TELQUAL_SEND, IAC, SE};
+    fd.write(buf);
+}
+
+void Channel::async_lookup_process(sockaddr_in address, Fd reply_fd) {
+    /*
+     * Firstly, and quite importantly, close all the descriptors we
+     * don't need, so we don't keep open sockets past their useful
+     * life.  This was a bug which meant ppl with firewalled auth
+     * ports could keep other ppls connections open (nonresponding)
+     */
+    doorman_->for_each_channel([](Channel &channel) { channel.close(); });
+
+    // Bail out on a PIPE - this means our parent has crashed
+    signal(SIGPIPE, [](int) {
+        log_out("hostname lookup process exiting on sigpipe");
+        exit(1);
+    });
+
+    auto *ent = gethostbyaddr((char *)&address.sin_addr, sizeof(address.sin_addr), AF_INET);
+
+    std::string hostName = ent ? ent->h_name : inet_ntoa(address.sin_addr);
+    size_t nBytes = hostName.size() + 1;
+    try {
+        reply_fd.write(nBytes);
+        reply_fd.write(hostName.c_str(), nBytes);
+    } catch (const std::runtime_error &re) {
+        log_out("ID: Unable to write to doorman - perhaps it crashed: %s", re.what());
+        return;
+    }
+    // Log the source hostname but mask it for privacy.
+    log_out("ID: Looked up %s", get_masked_hostname(hostName).c_str());
+}
+
+static bool SupportsAnsi(const char *src) {
+    static const char *terms[] = {"xterm",      "xterm-colour", "xterm-color", "ansi",
+                                  "xterm-ansi", "vt100",        "vt102",       "vt220"};
+    size_t i;
+    for (i = 0; i < (sizeof(terms) / sizeof(terms[0])); ++i)
+        if (!strcasecmp(src, terms[i]))
+            return true;
+    return false;
+}
+
+void Channel::closeConnection() {
+    // Log the source IP but masked for privacy.
+    log_out("[%d] Closing connection to %s", id_, get_masked_hostname(hostname_).c_str());
+    close();
+}
+
+void Channel::set_echo(bool echo) {
+    SendCom(fd_, echo ? WONT : WILL, TELOPT_ECHO);
+    echoing_ = echo;
+}
+
+void Channel::on_reconnect_attempt() {
+    if (!fd_.is_open())
+        return;
+    if (!sent_reconnect_message_) {
+        send_to_client("Attempting to reconnect to Xania"sv);
+        sent_reconnect_message_ = true;
+    }
+    send_to_client(".", 1);
+}
+
+void Channel::close() {
+    fd_.close();
+    host_lookup_fd_.close();
+    if (host_lookup_pid_) {
+        if (::kill(host_lookup_pid_, 9) < 0)
+            log_out("Couldn't kill child process (ident process has already exited)");
+        host_lookup_pid_ = 0;
+    }
+}
+
+void Channel::newConnection(Fd fd, sockaddr_in address) {
```

Comment:
nit: I notice some methods are snake case, others are camel, you were leaning towards snake right?

Reply:
Interesting,  Fd is being passed by value? And I see it uses std::move() to do so.

Reply:
> nit: I notice some methods are snake case, others are camel, you were leaning towards snake right?

Absolutely: I plan on a big rename as part of the "Coding convention renames"

>Interesting, Fd is being passed by value? And I see it uses std::move() to do so.

Yup! `Fd` is a "move only" type by dint of having its copy constructor (`Fd(const Fd&)`) and copy assign (`Fd &operator=(const Fd&)`) `delete`d. That means (this is non-normative, but mostly correctly) that to have an Fd you have to "own" it, either by having created it yourself, or having `std::move()`d it. That way we know with certainty that if you have an `Fd` you own it. If you let it fall out of scope, it will close. If you need to hand it to another function or object, you have to relinquish it and give it to that object (unless you use a temporary "lending" reference `const Fd &`, but I tried not to do that here).